coderd/userauth.go

```diff
@@ -28,8 +29,112 @@ import (
 	"github.com/coder/coder/coderd/rbac"
 	"github.com/coder/coder/coderd/userpassword"
 	"github.com/coder/coder/codersdk"
+	"github.com/coder/coder/cryptorand"
 )
 
+// postConvertLoginType replies with an oauth state token capable of converting
+// the user to an oauth user.
+func (api *API) postConvertLoginType(rw http.ResponseWriter, r *http.Request) {
```

Comment:
Do users need to be able to do this? What if we only allowed admins to change the state manually?

Reply:
User's trigger this. I took the idea from here https://github.com/coder/coder/issues/4505#issuecomment-1475658787


Essentially linking users to their OIDC account is tricky, as emails can match even if they are different people. Eg maybe an account was made with builtin called `paul@coder.com`. Paul leaves the company. The next Paul is hired and now it's difficult to know in the code if the Paul from OIDC is the same Paul from the builtin.

This resolves that by making it a user triggered action. The user goes to their Account > Settings. Hits "Convert to OIDC" and goes through the OIDC flow.

We generate a state string linking the OIDC login to the "Convert" request. So we know it is the same person as they were able to provide the password for the builtin AND provide the OIDC auth.

Reply:
Can we encode something in the `state` instead? We already sign the value, I believe.

Reply:
Err, if we don't we can. We have a global signing key that we use anyways.

Reply:
As for the why: I've always found it sketchy to insert any OAuth state into the database. It allows for an unauthenticated user to directly insert into the DB.

Reply:
I did think about signing state, but all Oauth 2.0 specs I read said it should be a non-guessable random string. I don't see a security risk with signing a random string and making it a JWT for example. We should have expiration and user_id as apart of the state string, so there is information that needs to be encoded.

Reply:
So some IDP's do have a maximum length state param. These values are different everywhere 

From https://auth0.com/docs/secure/attack-protection/state-parameters

> The allowed length for state is not unlimited. If you get the error 414 Request-URI Too Large, try a smaller value.

In practice, the smallest length I am finding is 255. It's not well documented anywhere.

[Bunq: Max length 255](https://together.bunq.com/d/7023-oauth-authorization-url-state-parameter-is-too-long)

Github, Gitlab, Google, etc do not document this max length. The common language is that the state string carries some CSRF protection. 

I did a test JWT and it was **332** characters long. 

----

In summary, I agree a JWT would be simpler as it requires no database interactions to pull out the info. And I honestly prefer to go the JWT method. My only fear is that someone using coder might hit a `state string too large` error with their IDP, and there would not be any recourse.

On the "Sketchy to insert OAuth sate into the database" because of unauthenticated users. In this case, only authenticated users can generate this `convert to OIDC` state string. An unauthenticated user cannot trigger this codepath.

I have code that makes sure only 1 state string per user exists in the DB.

So in this particular case, we avoid your worries.

If you still think we should go JWT instead, I can make the switch. I just honestly don't know what state string limitations exist, and I spent over an hour trying to find information without any luck besides errors people hit in random github issues on random projects.